sdk/test-utils/perfstress/src/testProxyHttpClient.ts

```diff
@@ -0,0 +1,272 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { HttpOperationResponse } from "@azure/core-http";
+import { DefaultHttpClient, WebResourceLike } from "@azure/core-http";
+import {
+  PipelinePolicy,
+  PipelineRequest,
+  PipelineResponse,
+  SendRequest
+} from "@azure/core-rest-pipeline";
+import { Dispatcher, request } from "undici";
+
+const paths = {
+  playback: "/playback",
+  record: "/record",
+  start: "/start",
+  stop: "/stop"
+};
+
+/**
+ * Helper class to manage the recording state to make sure the proxy-tool is not flooded with unintended requests.
+ *
+ * => then start record
+ * => run the runAsync
+ * => stop record
+ * => start playback
+ * => stop playback
+ */
+export class RecordingStateManager {
+  public state:
+    | "started-recording"
+    | "stopped-recording"
+    | "started-playback"
+    | "stopped-playback"
+    | undefined;
+
+  /**
+   * validateState
+   */
+  public validateState(
+    currentFlow:
+      | "starting-recording"
+      | "stopping-recording"
+      | "starting-playback"
+      | "stopping-playback"
+  ) {
+    if (currentFlow === "starting-recording") {
+      if (this.state === "started-recording") {
+        throw new Error("Already started recording, should not have called again.");
+      }
+    }
+    if (currentFlow === "stopping-recording") {
+      if (this.state === "stopped-recording") {
+        throw new Error("Already stopped recording, should not have called again.");
+      }
+      if (this.state !== "started-recording") {
+        throw new Error("Please start recording before calling stop.");
+      }
+    }
+    if (currentFlow === "starting-playback") {
+      if (this.state !== "stopped-recording") {
+        throw new Error("Did not stop recording, stop recording before starting playback.");
+      }
+    }
+    if (currentFlow === "stopping-playback") {
+      if (this.state !== "started-playback") {
+        throw new Error("Did not start playback, start playback before calling stop.");
+      }
+    }
+  }
+
+  /**
+   * setState
+   */
+  public setState(
+    state: "started-recording" | "stopped-recording" | "started-playback" | "stopped-playback"
+  ) {
+    this.state = state;
+  }
+}
+
+export class TestProxyHttpClientV1 extends DefaultHttpClient {
```

Comment:
why do you need to extend `DefaultHttpClient` here? It seems like it would be easier / safer to wrap it.

I actually wonder if it would be simpler to invert this and extend `TestProxyHttpClient` and then use a private `DefaultHttpClient` to actually do the request in the override of `sendRequest` ? You can still call `super` to do the transformation, right?

Reply:
I did think about it but I'm passing this as the httpClient in the client options, which requires a client of the same shape as HttpClient.
Now that I look more, all it needs is just the sendRequest.

I'll check and see if I can do the migration.

Reply:
This is perfect.
Very small code to maintain for core-v1 now.
Great feedback, Jeff. ❤